### PR TITLE
Filesapi renamefailcase

### LINKED ISSUE #26625: Node API's move doesn't check view->rename return value
### Steps

1. create a folder "test"
1. mount a system-wide ext storage on "/ext"
1. call `\OC::$server->getUserFolder()->get('test')->move('ext')

### Expected result
`move` operation fails with exception

### Actual result
`move` fails but the value of rename is not checked and the function returns the old node for "ext".

### Versions
Observed on master

Note: I didn't test the actual proposed case above yet but had a similar scenario
It is already clear from reading the code that the return value isn't checked: https://github.com/owncloud/core/blob/master/lib/private/Files/Node/Folder.php#L338

@owncloud/filesystem

### PR DESCRIPTION
**Pull request by PVince81**
## Description
- refactored Files Node API tests to have more coverage: basically make NodeTest abstract and put the common tests there
- refactored "rename()" and "copy()" operations to be in the Node class instead of File and Folder
- fixed "rename()" and "copy()" operations to not ignore failures

Why so much refatoring ? If I'd do the fix directly I'd have to add a lot of missing tests in the "FolderTest" class and also duplicate the fix across File and Folder.

Now I agree that throwing `NotPermittedException` in case of unknown failure from the view isn't very elegant. Any better suggestions welcome.

## Related Issue
Fixes https://github.com/owncloud/core/issues/26625

## Motivation and Context
Because we must not ignore errors.
Also, this blocks https://github.com/owncloud/core/issues/26615 where I need to rely on a move failure for a unit test.

## How Has This Been Tested?
Running unit tests

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Tech debt

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@jvillafanez @DeepDiver1975

### CODE REVIEW

File: lib/private/Files/Node/Node.php [hunk 1]
```diff
@@ -380,4 +366,54 @@ public function changeLock($type) {
...
+	 * @return \OC\Files\Node\Node
+	 */
+	public function move($targetPath) {
+		$targetPath = $this->normalizePath($targetPath);
+		$parent = $this->root->get(dirname($targetPath));
+		if ($parent instanceof Folder and $this->isValidPath($targetPath) and $parent->isCreatable()) {
+			$nonExisting = new $this->nonExistingClass($this->root, $this->view, $targetPath);
+			$this->root->emit('\OC\Files', 'preRename', [$this, $nonExisting]);
+			$this->root->emit('\OC\Files', 'preWrite', [$nonExisting]);
+			if (!$this->view->rename($this->path, $targetPath)) {
```
Comment (PVince81) [inline lib/private/Files/Node/Node.php]: this is the actual fix: add a `if` here and in the copy method later

File: lib/private/Files/Node/File.php [hunk 2]
```diff
@@ -28,6 +28,8 @@
 use OCP\Files\NotPermittedException;
 
 class File extends Node implements \OCP\Files\File {
+	protected $nonExistingClass = 'OC\Files\Node\NonExistingFile';
```
Comment (jvillafanez) [inline lib/private/Files/Node/File.php]: I think using a method is more clear. In the Node class you could use something like:
```
protected function getNonExistingClass() {
    if (__CLASS__ === 'OC\Files\Node\Node') {
        throw Exception();
    } else {
        $convertedClass = convertToNonExistingString(__CLASS__);  // 'OC\Files\Node\File' -> 'OC\Files\Node\NonExistingFile';
        return new $convertedClass();
    }
}
```
We'll need to define the expected return values for the function.

This way, the subclasses won't need to define the `$nonExistingClass` variable. As long as the method is aware that the target class might not exists, it should be fine

Comment (PVince81) [inline lib/private/Files/Node/File.php]: You're right. I was assuming that no one would ever bother to extend that class :smile: 

Will adjust...

**Comment by PVince81** [comment]
@jvillafanez adjusted slightly differently. I used a method to simply return the string name of the NonExisting class to use.

**Comment by jvillafanez** [comment]
Right now, the Node class needs to create the non-existent class, and to do that it needs to know how (particularly the parameters).

On the other hand, if an instance is returned, the Node class might only need to check if the instance is a Node class or subclass. In fact, taking into account the Node class isn't abstract we should make sure the returned instances are a subclass of Node but not Node.

Btw, the tests need to be adjusted.

**Comment by PVince81** [comment]
@jvillafanez I've now adjusted the node to have a method that creates the non existing instance for a given path.

The tests still pass and I think they are fine. I don't want to add the same logic there because the tests really only check instanceof so a string for the class name to verify is enough.

File: lib/private/Files/Node/File.php [hunk 3]
```diff
@@ -29,6 +29,16 @@
 
 class File extends Node implements \OCP\Files\File {
 	/**
+	 * Creates a Folder that represents a non-existing path
+	 *
+	 * @param string $path path
+	 * @return string non-existing node class
+	 */
+	protected function createNonExistingNode($path) {
+		$nonExisting = new NonExistingFile($this->root, $this->view, $path);
```
Comment (jvillafanez) [inline lib/private/Files/Node/File.php]: missing `return`

File: lib/private/Files/Node/Folder.php [hunk 4]
```diff
@@ -31,6 +31,16 @@
 
 class Folder extends Node implements \OCP\Files\Folder {
 	/**
+	 * Creates a Folder that represents a non-existing path
+	 *
+	 * @param string $path path
+	 * @return string non-existing node class
+	 */
+	protected function createNonExistingNode($path) {
+		$nonExisting = new NonExistingFolder($this->root, $this->view, $path);
```
Comment (jvillafanez) [inline lib/private/Files/Node/Folder.php]: missing `return`

File: lib/private/Files/Node/File.php (hunk 3 above)
Comment (PVince81) [inline lib/private/Files/Node/File.php]: uh oh... so the tests were lying... I'll add some hook tests to make sure this is caught too

**Comment by PVince81** [comment]
I fixed what you said and tried adding a test for the rename hook, but that one isn't done yet. Need to properly mock `$root` but keeping `listen` as is. Will need more work.

**Comment by PVince81** [comment]
So, added and fixed tests for the move/copy hooks.

Ready for another review.

**Comment by jvillafanez** [comment]
Code looks good 👍

**Pull request merged**